include/secp256k1.h

```diff
@@ -81,6 +82,29 @@ typedef struct {
     unsigned char data[64];
 } secp256k1_ecdsa_signature;
 
+/** Data structure that holds a sign-to-contract ("s2c") opening information.
+ *  Sign-to-contract allows a signer to commit to some data as part of a signature. It
+ *  can be used as an Out-argument in certain signing functions.
+ *
+ *  This structure is not opaque, but it is strongly discouraged to read or write to
+ *  it directly.
+ *
+ *  The exact representation of data inside is implementation defined and not
+ *  guaranteed to be portable between different platforms or versions. It can
+ *  be safely copied/moved.
+ */
+typedef struct {
+    /* magic is set during initialization */
+    uint64_t magic;
+    /* Public nonce before applying the sign-to-contract commitment */
+    secp256k1_pubkey original_pubnonce;
+    /* Byte indicating if signing algorithm negated the nonce. Alternatively when
+     * verifying we could compute the EC commitment of original_pubnonce and the
+     * data and negate if this would not be a valid nonce. But this would prevent
+     * batch verification of sign-to-contract commitments. */
+    int nonce_is_negated;
+} secp256k1_s2c_opening;
```

Comment:
My initial idea was that schnorr and ecdsa could reuse the same s2c structure, but I think it would simplify things quite a bit if they wouldn't. ECDSA doesn't make use of the `nonce_is_negated` flag. So we could make the object opaque and identical to pubkeys, delegate serialize and parse to pubkey ser and parse and move the opening struct out of the secp256k1 include and into secp256k1_s2c.h

Reply:
This sounds like a reasonable plan .

Reply:
Should we keep the magic field?

Reply:
No, calling `pubkey_load` is perfectly fine.